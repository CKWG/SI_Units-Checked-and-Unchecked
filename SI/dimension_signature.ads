------------------------------------------------------------------------------
-- Checked and Unchecked Computation with SI Units
-- Copyright (C) 2020 Christoph Karl Walter Grein
--
-- This program is free software; you can redistribute it and/or
-- modify it under the terms of the GNU General Public License
-- as published by the Free Software Foundation; either version 3
-- of the License, or any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program; if not, write to the Free Software
-- Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
--
-- As a special exception,  if other files  instantiate  generics from this
-- unit, or you link this unit with other files to produce an executable,
-- this unit does not by itself cause the resulting executable to be
-- covered by the GNU General Public License.  This exception does not
-- however invalidate any other reasons why the executable file might be
-- covered by the GNU Public License.
--
-- Source:
-- https://www.adaic.org/ada-resources/tools-libraries/
--   (see Christoph Grein's Essentials)
-- http://archive.adaic.com/tools/CKWG/Dimension/Dimension.html
--
-- Author's email address:
--   christ-Usch.grein@t-online.de
------------------------------------------------------------------------------

with Rational_Arithmetics;
use  Rational_Arithmetics;

generic

  --====================================================================
  -- Author    Christoph Grein
  -- Version   1.0
  -- Date      11 May 2020
  --====================================================================
  -- Signature package for Generic_SI defining the dimension handling.
  --====================================================================
  -- History
  -- Author Version   Date    Reason for change
  --  C.G.    1.0  11.05.2020
  --====================================================================

  type Dimension is private;

  uno: Dimension;

  with function m   return Dimension is <>;
  with function kg  return Dimension is <>;
  with function s   return Dimension is <>;
  with function A   return Dimension is <>;
  with function K   return Dimension is <>;
  with function cd  return Dimension is <>;
  with function mol return Dimension is <>;

  with function "*" (Left, Right: Dimension) return Dimension is <>;
  with function "/" (Left, Right: Dimension) return Dimension is <>;

  with function "**" (Base: Dimension; Exponent: Whole   ) return Dimension is <>;
  with function "**" (Base: Dimension; Exponent: Rational) return Dimension is <>;

  -- Access to the components

  with function m   (X: Dimension) return Rational is <>;
  with function kg  (X: Dimension) return Rational is <>;
  with function s   (X: Dimension) return Rational is <>;
  with function A   (X: Dimension) return Rational is <>;
  with function K   (X: Dimension) return Rational is <>;
  with function cd  (X: Dimension) return Rational is <>;
  with function mol (X: Dimension) return Rational is <>;

package Dimension_Signature is
end Dimension_Signature;
