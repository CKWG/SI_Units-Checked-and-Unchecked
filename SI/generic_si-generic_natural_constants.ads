------------------------------------------------------------------------------
-- Checked and Unchecked Computation with SI Units
-- Copyright (C) 2002, 2005, 2006, 2018, 2019, 2020
--               Christoph Karl Walter Grein
--
-- This program is free software; you can redistribute it and/or
-- modify it under the terms of the GNU General Public License
-- as published by the Free Software Foundation; either version 3
-- of the License, or any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program; if not, write to the Free Software
-- Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
--
-- As a special exception, if other files instantiate generics from this
-- unit, or you link this unit with other files to produce an executable,
-- this unit does not by itself cause the resulting executable to be
-- covered by the GNU General Public License. This exception does not
-- however invalidate any other reasons why the executable file might be
-- covered by the GNU Public License.
--
-- Author's homepage and email address:
--   http://archive.adaic.com/tools/CKWG/CKWG.html
--   Christ-Usch.Grein@T-Online.de
------------------------------------------------------------------------------

generic

package Generic_SI.Generic_Natural_Constants is

  --====================================================================
  -- Author    Christoph Grein
  -- Version   4.1
  -- Date      9 August 2020
  --====================================================================
  -- SI Redefinition became effective on May 20, 2019.
  -- Didn't find a better abbreviation for e_e, normally named just e.
  -- e_e stands for e(elementary).
  -- Change values, if necessary, to the newest measured values.
  --====================================================================
  -- History
  -- Author Version   Date    Reason for change
  --  C.G.    0.0  03.09.2002 Only sketched
  --  C.G.    1.0  23.09.2005 Added some important constants
  --  C.G.    1.1  26.02.2006 pragma Pure
  --  C.G.    2.0  23.07.2018 Unit strings
  --  C.G.    2.1  28.09.2018 Unit syntax changed W/m/K => W/(m*K)
  --  C.G.    3.0  02.10.2018 Newest constant values from NIST
  --  C.G.    3.1  20.05.2019 World Metrology Day
  --  C.G.    3.2  07.03.2020 Stefan-Boltzmann and Gas constants
  --  C.G.    4.0  11.05.2020 Parent renamed to Generic_SI
  --  C.G.    4.1  09.08.2020 Name Speed_of_Light added
  --====================================================================

  -- Speed of Light in Vacuum

  Speed_of_Light: constant Speed := 2.997_924_58E+8*"m/s";  -- exact (by definition)
  c: Speed renames Speed_of_Light;

  -- Particles

  Elementary_Charge: constant Charge := 1.602_176_634E-19*"C";  -- exact (by definition)
  e_e: Charge renames Elementary_Charge;

  Electron_Mass: constant Mass := 9.109_382_6E-31 *"kg";
  Proton_Mass  : constant Mass := 1.672_621_71E-27*"kg";

  m_e: Mass renames Electron_Mass;
  m_p: Mass renames Proton_Mass;

  -- Permittivity and Permeability of Vacuum

  Mu_0 : constant Item := 4.0E-7*Pi*"kg*m/C**2";  -- exact; c = 1/Sqrt(Eps_0*Mu_0)
  Eps_0: constant Item := 1.0/(Mu_0*c**2);        -- 8.854_187_817E-12 F/m

  -- Quantum Mechanics

  Planck: constant Item := 6.626_070_15E-34*"J*s";  -- exact (by definition)
  h     :          Item renames Planck;
  h_bar : constant Item := h / (2.0*Pi);

  Alpha: constant Dimensionless := e_e**2/(2.0*Eps_0*h*c);  -- 1/137.035_999_11

  -- Thermodynamics

  Boltzmann: constant Item := 1.380_649E-23*"J/K";     -- exact (by definition)
  Avogadro : constant Item := 6.022_140_76E+23/"mol";  -- exact (by definition)

  k_B: Item renames Boltzmann;
  N_A: Item renames Avogadro;

  Gas_Constant: constant Item := N_A * k_B;  -- exact; 8.314 J/(mol*K)

  -- Black body radiation

  Stefan_Boltzmann: constant Item := (2.0*Pi**5/15.0)*(k_B/h)**3*(k_B/c**2);  -- careful with underflow
  Sigma           :          Item renames Stefan_Boltzmann;  -- exact; 5.670E-8*W/(m**2*K**4)

  -- Gravity

  Gravity: constant Item := 6.674_2E-11*"m**3/(kg*s**2)";

end Generic_SI.Generic_Natural_Constants;
