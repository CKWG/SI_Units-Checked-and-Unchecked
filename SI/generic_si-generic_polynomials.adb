------------------------------------------------------------------------------
-- Checked and Unchecked Computation with SI Units
-- Copyright (C) 2002, 2018, 2020 Christoph Karl Walter Grein
--
-- This program is free software; you can redistribute it and/or
-- modify it under the terms of the GNU General Public License
-- as published by the Free Software Foundation; either version 3
-- of the License, or any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program; if not, write to the Free Software
-- Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
--
-- As a special exception,  if other files  instantiate  generics from this
-- unit, or you link this unit with other files to produce an executable,
-- this unit does not by itself cause the resulting executable to be
-- covered by the GNU General Public License.  This exception does not
-- however invalidate any other reasons why the executable file might be
-- covered by the GNU Public License.
--
-- Source:
-- https://www.adaic.org/ada-resources/tools-libraries/
--   (see Christoph Grein's Essentials)
-- http://archive.adaic.com/tools/CKWG/Dimension/Dimension.html
--
-- Author's email address:
--   christ-usch.grein@t-online.de
------------------------------------------------------------------------------

package body Generic_SI.Generic_Polynomials is

  --====================================================================
  -- Author    Christoph Grein
  -- Version   5.0
  -- Date      14 May 2020
  --====================================================================
  --
  --====================================================================
  -- History
  -- Author Version   Date    Reason for change
  --  C.G.    1.0  07.12.1994
  --  C.G.    1.1  13.05.1997 Interface changed
  --  C.G.    2.0  31.07.1997 Interface rewrought
  --  C.G.    3.0  26.07.2002 Adapted to SI package
  --  C.G.    4.0  26.08.2018 Unit strings
  --  C.G.    5.0  14.05.2020 Dimensions generic parameter
  --====================================================================

  function Polynomial (Coeff: Coefficients; X: Item) return Item is
    -- Use Horner's scheme.
    F: Item := (Value => 0.0, Unit => Coeff (Coeff'Last).Unit / X.Unit);
  begin
    for I in reverse Coeff'Range loop
      F := F * X + Coeff (I);
    end loop;
    return F * X ** Coeff'First;
  end Polynomial;

  function Linear_Regression (S: Sample) return Straight_Line is
    Sx : Item := 0.0 * S (S'First).X;
    Sy : Item := 0.0 * S (S'First).Y;                  -- Get
    Sxy: Item := 0.0 * S (S'First).X * S (S'First).Y;  -- the
    Sx2: Item := 0.0 * S (S'First).X**2;               -- unit
  begin
    if S'Length <= 1 then
      raise Sample_Error;
    end if;
    for Si of S loop
      Sx  := Sx  + Si.X;
      Sy  := Sy  + Si.Y;
      Sxy := Sxy + Si.X * Si.Y;
      Sx2 := Sx2 + Si.X**2;
    end loop;
    declare
      N: constant Real := Real (S'Length);
      M: constant Item := (N * Sxy - Sx * Sy) / (N * Sx2 - Sx**2);
    begin
      return (0 => (Sy - M * Sx) / N,
              1 => M);
    end;
  end Linear_Regression;

  function Define (S: Sample) return Interpolation_Table is
    -- Check conditions on input tables.
    -- Make the internal table increasing and store direction in sign:
    --   Sign := +1 (increasing);
    --   Sign :=  0 (table error);
    --   Sign := -1 (decreasing).
    -- Liniear interpolation will be called with signed argument "Sign * X".
    Table: Interpolation_Table (S'Length);
  begin
    if S'Length <= 1 then
      raise Sample_Error;
    end if;
    -- Check monotony
    if S (S'First).X < S (S'First + 1).X then
      Table.Sign := +1.0;  -- increasing
    else
      Table.Sign := -1.0;  -- decreasing
    end if;
    Table.S (1) := (X => Table.Sign * S (S'First).X,
                    Y =>              S (S'First).Y);
    for I in 2 .. Table.Length loop
      Table.S (I) := (X => Table.Sign * S (S'First + I - 1).X,
                      Y =>              S (S'First + I - 1).Y);
      if Table.S (I - 1).X >= Table.S (I).X then
        raise Sample_Error;
      end if;
    end loop;
    return Table;
  end Define;

  function Linear_Interpolation
               (Table: Interpolation_Table; X: Item) return Item is
    -- Check whether table is defined.
    -- Use a binary search to find X's position in the table.
    Signed_X: constant Item := Table.Sign * X;
    Lower, Upper: Positive;
    function Binary_Search return Positive is
      L: Positive := 1;
      U: Positive := Table.Length;
      M: Positive;
    begin
      while U - L > 1 loop
        M := Integer (L + U) / 2;
        if Signed_X <= Table.S (M).X then
          U := M;
        else
          L := M;
        end if;
      end loop;
      return L;
    end Binary_Search;
  begin
    if Table.Sign = 0.0 then
      raise Sample_Error;
    end if;
    if Signed_X < Table.S (1).X then
      Lower := 1;
    elsif Signed_X > Table.S (Table.Length).X then
      Lower := Table.Length - 1;
    else
      Lower := Binary_Search;
    end if;
    Upper := Lower + 1;
    return  Table.S (Lower).Y +
           (Table.S (Upper).Y - Table.S (Lower).Y) * ((         Signed_X - Table.S (Lower).X) /
                                                      (Table.S (Upper).X - Table.S (Lower).X));
  end Linear_Interpolation;

end Generic_SI.Generic_Polynomials;
