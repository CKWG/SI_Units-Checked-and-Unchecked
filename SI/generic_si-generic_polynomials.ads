------------------------------------------------------------------------------
-- Checked and Unchecked Computation with SI Units
-- Copyright (C) 2002, 2006, 2018, 2020 Christoph Karl Walter Grein
--
-- This program is free software; you can redistribute it and/or
-- modify it under the terms of the GNU General Public License
-- as published by the Free Software Foundation; either version 3
-- of the License, or any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program; if not, write to the Free Software
-- Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
--
-- As a special exception,  if other files  instantiate  generics from this
-- unit, or you link this unit with other files to produce an executable,
-- this unit does not by itself cause the resulting executable to be
-- covered by the GNU General Public License.  This exception does not
-- however invalidate any other reasons why the executable file might be
-- covered by the GNU Public License.
--
-- Source:
-- https://www.adaic.org/ada-resources/tools-libraries/
--   (see Christoph Grein's Essentials)
-- http://archive.adaic.com/tools/CKWG/Dimension/Dimension.html
--
-- Author's email address:
--   christ-usch.grein@t-online.de
------------------------------------------------------------------------------

generic
package Generic_SI.Generic_Polynomials is

  --====================================================================
  -- Author    Christoph Grein
  -- Version   5.0
  -- Date      14 May 2020
  --====================================================================
  -- Polynomial defines the function
  --    f(x) := sum (Coeff (i) * X**i)
  -- Coefficients' range need not include 0, e.g.
  --    Coefficients = (1=> 1.0, 2=> 1.0)
  -- defines the term x*(x+1).
  --
  -- Linear_Regression defines the coefficients for a linear function
  -- approximating a sample of points. The sample S must at least hold
  -- two points, else Sample_Error is raised.
  --
  -- Define creates a table for linear interpolation. The sample S must
  -- at least hold two points, else Sample_Error is raised. The length
  -- of the table returned is S'Length.
  -- X and Y denote the abscissa and ordinate values of sample points.
  -- X must be strictly monotonic (increasing or decreasing) or else
  -- Sample_Error is raised.
  -- Linear_Interpolation interpolates linearly for x within the Table
  -- range and extrapolates outside.
  -- Table_Error is propagated if the table is undefined.
  --====================================================================
  -- History
  -- Author Version   Date    Reason for change
  --  C.G.    1.0  07.12.1994
  --  C.G.    1.1  13.05.1997 Interface changed
  --  C.G.    2.0  31.07.1997 Interpolation table made private
  --  C.G.    3.0  26.07.2002 Adapted to SI package
  --  C.G.    3.1  26.02.2006 pragma Pure
  --  C.G.    4.0  26.08.2018 Unit strings
  --  C.G.    5.0  14.05.2020 Dimensions generic parameter
  --====================================================================

  -- Polynomials

  type Coefficients is array (Natural_Whole range <>) of Item;
  subtype Straight_Line is Coefficients (0 .. 1);

  function Polynomial (Coeff: Coefficients; X: Item) return Item;

  -- Linear Regression and Interpolation

  type Pair is record
    X, Y: Item;  -- y=f(X)
  end record;

  type Sample is array (Positive range <>) of Pair;

  function Linear_Regression (S: Sample) return Straight_Line;

  type Interpolation_Table (Length: Positive) is private;
  function Define (S: Sample) return Interpolation_Table;

  function Linear_Interpolation
               (Table: Interpolation_Table; X: Item) return Item;

  Sample_Error: exception;

private

  type Interpolation_Table (Length: Positive) is record
    Sign: Real := 0.0;  -- means undefined
    S: Sample (1 .. Length);
  end record;

end Generic_SI.Generic_Polynomials;
