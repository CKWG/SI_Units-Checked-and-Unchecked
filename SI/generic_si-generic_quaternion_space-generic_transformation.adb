------------------------------------------------------------------------------
-- Checked and Unchecked Computation with SI Units
-- Copyright (C) 2004 Chris Holmes
-- Copyright (C) 2004, 2018, 2020 Christoph Grein
--
-- This program is free software; you can redistribute it and/or
-- modify it under the terms of the GNU General Public License
-- as published by the Free Software Foundation; either version 3
-- of the License, or any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program; if not, write to the Free Software
-- Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
--
-- As a special exception,  if other files  instantiate  generics from this
-- unit, or you link this unit with other files to produce an executable,
-- this unit does not by itself cause the resulting executable to be
-- covered by the GNU General Public License.  This exception does not
-- however invalidate any other reasons why the executable file might be
-- covered by the GNU Public License.
--
-- Source:
-- https://www.adaic.org/ada-resources/tools-libraries/
--   (see Christoph Grein's Essentials)
-- http://archive.adaic.com/tools/CKWG/Dimension/Dimension.html
--
-- Author's email address:
--   christ-Usch.grein@t-online.de
------------------------------------------------------------------------------

package body Generic_SI.Generic_Quaternion_Space.Generic_Transformation is

  --====================================================================
  -- Authors   Chris Holmes, Christoph Grein
  -- Version   3.0
  -- Date      15 May 2020
  --====================================================================
  --
  --====================================================================
  -- History
  -- Author Version   Date    Reason for change
  --  CKH     0.0  18.03.2004 Prototype for rotation around the three
  --                          coordinate axes
  --  C.G.    1.0  10.05.2004 Function wrapped in child package
  --  C.G.    1.1  06.09.2004 Rotation around an axis added
  --  C.G.    2.0  04.08.2018 Unit strings
  --  C.G.    3.0  15.05.2020 Dimensions generic parameter
  --====================================================================

  function Rotation (Axis: Vector; By: Angle) return Quaternion is
    A: constant Angle := By/2.0;
  begin
    return (Re => Cos (A), Im => Sin (A) * Axis / abs Axis);
  end Rotation;

  function Rotation (Yaw, Pitch, Roll: Angle := 0.0 * "rad") return Quaternion is
    -- Q = (C3, (0, 0, S3)) * (C2, (0, S2, 0)) * (C1, (S1, 0, 0))
    R : constant Angle := Roll /2.0;
    P : constant Angle := Pitch/2.0;
    Y : constant Angle := Yaw  /2.0;
    C1: constant Item := Cos (R);
    S1: constant Item := Sin (R);
    C2: constant Item := Cos (P);
    S2: constant Item := Sin (P);
    C3: constant Item := Cos (Y);
    S3: constant Item := Sin (Y);
  begin
    return (Re =>                C1*C2*C3 + S1*S2*S3,
            Im => Compose (V1 => S1*C2*C3 - C1*S2*S3,
                           V2 => C1*S2*C3 + S1*C2*S3,
                           V3 => C1*C2*S3 - S1*S2*C3));
  end Rotation;

end Generic_SI.Generic_Quaternion_Space.Generic_Transformation;
