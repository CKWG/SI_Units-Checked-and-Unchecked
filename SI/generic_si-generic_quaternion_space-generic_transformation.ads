------------------------------------------------------------------------------
-- Checked and Unchecked Computation with SI Units
-- Copyright (C) 2004, 2006, 2018, 2020 Christoph Grein
--
-- This program is free software; you can redistribute it and/or
-- modify it under the terms of the GNU General Public License
-- as published by the Free Software Foundation; either version 3
-- of the License, or any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program; if not, write to the Free Software
-- Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
--
-- As a special exception,  if other files  instantiate  generics from this
-- unit, or you link this unit with other files to produce an executable,
-- this unit does not by itself cause the resulting executable to be
-- covered by the GNU General Public License.  This exception does not
-- however invalidate any other reasons why the executable file might be
-- covered by the GNU Public License.
--
-- Source:
-- https://www.adaic.org/ada-resources/tools-libraries/
--   (see Christoph Grein's Essentials)
-- http://archive.adaic.com/tools/CKWG/Dimension/Dimension.html
--
-- Author's email address:
--   christ-Usch.grein@t-online.de
------------------------------------------------------------------------------

generic
package Generic_SI.Generic_Quaternion_Space.Generic_Transformation is

  --====================================================================
  -- Authors   Christoph Grein
  -- Version   3.0
  -- Date      15 May 2020
  --====================================================================
  -- A rotation of a vector V via a quaternion Q is achieved by trans-
  -- forming V with the following equation:
  --
  --   VQ' := Q * VQ / Q
  --
  -- or for a normalised quaternion (1/Q = Conjugate (Q)):
  --
  --   VQ' := Q * VQ * Conjugate (Q)
  --
  -- where Re (VQ) = 0, Im (VQ) = V, and the rotated vector V' is the
  -- imaginary part of VQ':
  --
  --   V' = Im (VQ')
  --
  -- Note that the vector is rotated, not the coordinate system.
  --
  -- The functions provided deliver normalised quaternions.
  --
  -- 1. Rotation (right-handed) around an axis given by a vector.
  --
  --    function Rotation (Axis: Vector; By: Angle) return Quaternion;
  --
  -- 2. Successive rotation (right-handed) around the three coordinate
  --    axes.
  --
  --    function Rotation (Yaw, Pitch, Roll: Angle := 0.0 * "rad")
  --      return Quaternion;
  --
  --    Roll : around x-axis (first rotation)
  --    Pitch: around y-axis (second rotation)
  --    Yaw  : around z-axis (third rotation)
  --
  --    This transformation Q fulfills the equation
  --
  --    Q (Yaw, Pitch, Roll) = Q (Yaw) * Q (Pitch) * Q (Roll)
  --====================================================================
  -- History
  -- Author Version   Date    Reason for change
  --  C.G.    1.0  10.05.2004
  --  C.G.    1.1  20.07.2004 Rotation around an axis added
  --  C.G.    1.2  26.02.2006 pragma Pure
  --  C.G.    2.0  04.08.2018 Unit strings
  --  C.G.    3.0  15.05.2020 Dimensions generic parameter
  --====================================================================

  use Vectors;

  function Rotation (Axis: Vector; By: Angle) return Quaternion;

  function Rotation (Yaw, Pitch, Roll: Angle := 0.0 * "rad") return Quaternion;

end Generic_SI.Generic_Quaternion_Space.Generic_Transformation;
