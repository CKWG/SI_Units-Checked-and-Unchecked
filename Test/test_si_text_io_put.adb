------------------------------------------------------------------------------
-- Checked and Unchecked Computation with SI Units
-- Copyright (C) 2002, 2018, 2020 Christoph Karl Walter Grein
--
-- This program is free software; you can redistribute it and/or
-- modify it under the terms of the GNU General Public License
-- as published by the Free Software Foundation; either version 3
-- of the License, or any later version.
--
-- This program is distributed in the hope that it will be useful,
-- but WITHOUT ANY WARRANTY; without even the implied warranty of
-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-- GNU General Public License for more details.
--
-- You should have received a copy of the GNU General Public License
-- along with this program; if not, write to the Free Software
-- Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
--
-- As a special exception,  if other files  instantiate  generics from this
-- unit, or you link this unit with other files to produce an executable,
-- this unit does not by itself cause the resulting executable to be
-- covered by the GNU General Public License.  This exception does not
-- however invalidate any other reasons why the executable file might be
-- covered by the GNU Public License.
--
-- Source:
-- https://www.adaic.org/ada-resources/tools-libraries/
--   (see Christoph Grein's Essentials)
-- http://archive.adaic.com/tools/CKWG/Dimension/Dimension.html
--
-- Author's email address:
--   christ-usch.grein@t-online.de
------------------------------------------------------------------------------

with Ada.Text_IO, Ada.Float_Text_IO;

with Test_Support;
use  Test_Support;

with SI.IO;
use  SI.IO, SI;
with Ada.Assertions;

procedure Test_SI_Text_IO_Put is

  --====================================================================
  -- Author    Christoph Grein
  -- Version   3.1
  -- Date      3 July 2020
  --====================================================================
  -- Test the Fore, Exp, Dim format modifiers of Put.
  -- Compare the output in file Test_SI_Text_IO_Put.out with
  -- that in file Test_SI_Text_IO_Put.out.expected.
  -- Note: The test cannot be run with the unchecked version since
  --       dimension output is vital.
  --====================================================================
  -- History
  -- Author Version   Date    Reason for change
  --  C.G.    1.0  06.08.2002
  --  C.G.    2.0  03.08.2018 Unit strings
  --  C.G.    2.1  29.09.2018 Unit syntax changed
  --  C.G.    3.0  14.05.2020 Dimensions generic parameter
  --  C.G.    3.1  03.07.2020 Split a test string into two
  --====================================================================

  package A renames Ada.Text_IO;

  Physic, Result: A.File_Type;

begin

  Test_Header (Title => "Test SI.Put",
               Description => "Test formatting parameters.");

  Test_Step (Title => "Assertions",
             Description => "Test that Assertion_Policy is Check.");

  declare
    T: Time;
  begin
    T := 1.0*"S";
    Assert (Condition => False,
            Message   => "Switch on Assertion_Policy and use checked instantiation",
            Only_Report_Error => False);
    Test_Result;
    return;
  exception
    when Ada.Assertions.Assertion_Error =>
      Assert (Condition => True,
              Message   => "Exception Assertion_Error as expected",
              Only_Report_Error => True);
  end;

  -----------------------------------------------------------

  Test_Step (Title => "Compare",
             Description => "Compare the actual with the expected output.");

  A.Create (Result, A.Out_File, "Test_SI_Text_IO_Put.out");

  A.Set_Output (Result);

  Put (1.0 * "km/s");                          New_Line;
  Put (1.0 * "km/s", Dim => "*km/s");          New_Line;
  Put (1.0 * "km/s", Dim => "/(ms*m**(-1))");  New_Line;

  begin
    Put (1.0 * "km/s", Dim => "Km/s");
  exception
    when Illegal_Unit => Put_Line ("Illegal_Unit wrong casing Dim Km/s");
  end;

  begin
    Put (1.0 * "Km/s", Dim => "km/s");
  exception
    when Illegal_Unit => Put_Line ("Illegal_Unit wrong casing unit Km/s");
  end;

  begin
    Put (1.0 * "km/s", Dim => "km/s   A");
  exception
    when Illegal_Unit => Put_Line ("Illegal_Unit wrong format km/s   A");
  end;

  begin
    Put (1.0 * "s", Dim => "S");
  exception
    when Unit_Error => Put_Line ("Unit_Error wrong unit S (must be s)");
  end;

  -- Currently illegal, might be implemented some day
  -- Put (1.0 * "(km/s)**2", Dim => "km**2/s**2");  New_Line;
  -- Put (1.0 * "km/s"**2  , Dim => "(km/s)**2" );  New_Line;  Bad idea?

  New_Line;
  Put_Line ("Compare Ada.Text_IO.Float_Text_IO with SI.Text_IO;");
  Put_Line ("number formats must be identical:");
  New_Line;

  A.Put ('|');  Ada.Float_Text_IO.Put (1000.0, Fore => 0, Aft => 2, Exp => 0);  A.Put ('|');  A.Set_Col (15);  A.Put ('|');  Put (1.0 * "km/s", Fore => 0, Aft => 2, Exp => 0);  A.Put ('|');  New_Line;
  A.Put ('|');  Ada.Float_Text_IO.Put (1000.0, Fore => 0, Aft => 2, Exp => 1);  A.Put ('|');  A.Set_Col (15);  A.Put ('|');  Put (1.0 * "km/s", Fore => 0, Aft => 2, Exp => 1);  A.Put ('|');  New_Line;
  A.Put ('|');  Ada.Float_Text_IO.Put (1000.0, Fore => 0, Aft => 2, Exp => 3);  A.Put ('|');  A.Set_Col (15);  A.Put ('|');  Put (1.0 * "km/s", Fore => 0, Aft => 2, Exp => 3);  A.Put ('|');  New_Line;
  A.Put ('|');  Ada.Float_Text_IO.Put (1000.0, Fore => 3, Aft => 2, Exp => 0);  A.Put ('|');  A.Set_Col (15);  A.Put ('|');  Put (1.0 * "km/s", Fore => 3, Aft => 2, Exp => 0);  A.Put ('|');  New_Line;
  A.Put ('|');  Ada.Float_Text_IO.Put (1000.0, Fore => 3, Aft => 2, Exp => 1);  A.Put ('|');  A.Set_Col (15);  A.Put ('|');  Put (1.0 * "km/s", Fore => 3, Aft => 2, Exp => 1);  A.Put ('|');  New_Line;
  A.Put ('|');  Ada.Float_Text_IO.Put (1000.0, Fore => 3, Aft => 2, Exp => 3);  A.Put ('|');  A.Set_Col (15);  A.Put ('|');  Put (1.0 * "km/s", Fore => 3, Aft => 2, Exp => 3);  A.Put ('|');  New_Line;
  A.Put ('|');  Ada.Float_Text_IO.Put (1000.0, Fore => 5, Aft => 2, Exp => 0);  A.Put ('|');  A.Set_Col (15);  A.Put ('|');  Put (1.0 * "km/s", Fore => 5, Aft => 2, Exp => 0);  A.Put ('|');  New_Line;
  A.Put ('|');  Ada.Float_Text_IO.Put (1000.0, Fore => 5, Aft => 2, Exp => 1);  A.Put ('|');  A.Set_Col (15);  A.Put ('|');  Put (1.0 * "km/s", Fore => 5, Aft => 2, Exp => 1);  A.Put ('|');  New_Line;
  A.Put ('|');  Ada.Float_Text_IO.Put (1000.0, Fore => 5, Aft => 2, Exp => 3);  A.Put ('|');  A.Set_Col (15);  A.Put ('|');  Put (1.0 * "km/s", Fore => 5, Aft => 2, Exp => 3);  A.Put ('|');  New_Line;

  A.Set_Output (A.Standard_Output);  -- else Reset raises Mode_Error
  A.Reset (Result, A.In_File);
  A.Open  (Physic, A.In_File , "Test_SI_Text_IO_Put.out.expected");

  declare
    Line_P, Line_R: String (1 .. 55);
    Last_P, Last_R: Natural;
  begin
    loop
      A.Get_Line (Physic, Line_P, Last_P);
      A.Get_Line (Result, Line_R, Last_R);
      Assert (Condition => Line_P (1 .. Last_P) = Line_R (1 .. Last_R),
              Message   => Line_R (1 .. Last_R),
              Only_Report_Error => False);
    end loop;
  exception
    When A.End_Error => null;
  end;

  A.Close  (Physic);
  A.Delete (Result);

  Test_Result;

end Test_SI_Text_IO_Put;
