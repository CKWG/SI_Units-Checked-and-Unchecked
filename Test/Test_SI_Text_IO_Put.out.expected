 1.00000E+03*m*s**(-1)
 1.00000E+00*km/s
 1.00000E+00/(ms*m**(-1))
Illegal_Unit wrong casing Dim Km/s
Illegal_Unit wrong casing unit Km/s
Illegal_Unit wrong format km/s   A
Unit_Error wrong unit S (must be s)

Compare Ada.Text_IO.Float_Text_IO with SI.Text_IO;
number formats must be identical:

|1000.00|     |1000.00*m*s**(-1)|
|1.00E+3|     |1.00E+3*m*s**(-1)|
|1.00E+03|    |1.00E+03*m*s**(-1)|
|1000.00|     |1000.00*m*s**(-1)|
|  1.00E+3|   |  1.00E+3*m*s**(-1)|
|  1.00E+03|  |  1.00E+03*m*s**(-1)|
| 1000.00|    | 1000.00*m*s**(-1)|
|    1.00E+3| |    1.00E+3*m*s**(-1)|
|    1.00E+03||    1.00E+03*m*s**(-1)|
